### DISCUSSION TITLE
I2S to a2dp a2dpStream   . . . waiting for buffer to be consumed

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#923] [Q&A] [answered]
I2s to a2dpStream I'm getting waiting for buffer to be consumed. What this mean ? What the solution ? 

I2S rx at (  44100 2 16 )   

11:00:38.150 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.150 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.150 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.150 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.150 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.197 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.197 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.197 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.197 -> [I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
11:00:38.197 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...
11:00:38.427 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...
11:00:38.612 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...
11:00:38.796 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...
11:00:39.027 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...
11:00:39.214 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...
11:00:39.399 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...
11:00:39.629 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...
11:00:39.814 -> [I] AudioA2DP.h : 197 - waiting for buffer to be consumed...

**Comment by pschatzmann** [comment]
Which sketch are you trying to use ? 
Did you forget to call begin() ?
What exactly are you trying to do ?

Write is filling a buffer which needs to be consumed by A2DP callbacks.
I would recommend to avoid the A2DPStream and use the basic examples instead: they are much more efficient.

**Comment by aqildad-create** [comment]
**Which sketch are you trying to use ?** I  just slightly modify i2s to a2dp 
 **Did you forget to call begin() ?** I thing begin is their  that why    Its coping for some time then give message  buffer to be consume then then again coping for some time .. you can confrm this from above log .. 
**What exactly are you trying to do ?** read i2s and send to a2dp . 

**Write is filling a buffer which needs to be consumed by A2DP callbacks.** why is this am i coping fast ?I  use ( 44100 2 16 )  because  thought a2dp support this right ? 

**I would recommend to avoid the A2DPStream and use the basic examples instead: they are much more efficient.**
how to send live 2s data using basic examples ?

**Comment by pschatzmann** [comment]
This is just an information: if you want the message to go away, just change the log level to warning.

**Reply by aqildad-create** [reply]
I don't want the message to go way 
 
 my concern is how to handle this ?  I want to send continuous data from I2s  . . Is this bug should we convert this discussion to issue ?

**Reply by pschatzmann** [reply]
Information messages are not bugs: just information!

**Comment by aqildad-create** [comment]
problem is not with the information . 
 
  why the buffer is filled ? If feeding data at proper rate (44100 )   ??? I am stuck ?

**Comment by pschatzmann** [comment]
As I said, everything works as designed: The copy just tries to keep the buffer full and if it is already full it is waiting for a now buffer to be available, issuing this message...

**Reply by aqildad-create** [reply]
So this is OK that it take approx. 10,15 seconds to buffer full and sending together while  then take 5,10 seconds buffer to available again ?

 by this how can i send continuous i2s data ?? 

does it has link with distance b/w two esp32 ? if distance increase then the connection may break right ?? Why this buffer wait for availability is slow ? It take 5,10 seconds buffer to available again

**Reply by pschatzmann** [reply, marked as answer]
You mean millisecondss ? It can't be 15 seconds!
In any case, as I already told you above you should avoid to use the A2DPStream and use the example from the basic examples instead!
This does not need any buffer at all!

**Reply by aqildad-create** [reply]
Its seconds not milliseconds .

In basic example how to read i2s and send to a2dp ? 

you can see below more then 1.5 seconds has passed in below logs for buffer availability 

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/65456998/658c4bac-0b22-43f0-bd7b-e0254ef41e74)
